TestNodeDependOn is not stable.

### Which jobs are flaking?

https://github.com/pingcap/tiflow/pull/6219 leak_test 

### Which test(s) are flaking?

TestNodeDependOn

### Jenkins logs or GitHub Actions link

```log
==================

[2022-07-08T04:47:41.116Z] WARNING: DATA RACE

[2022-07-08T04:47:41.116Z] Write at 0x00c0001aac80 by goroutine 12:

[2022-07-08T04:47:41.116Z]   github.com/pingcap/tiflow/pkg/causality/internal.NewNode.func1()

[2022-07-08T04:47:41.116Z]       /home/jenkins/agent/workspace/cdc_ghpr_leak_test/go/src/github.com/pingcap/tiflow/pkg/causality/internal/node.go:77 +0x6c

[2022-07-08T04:47:41.116Z]   runtime.deferreturn()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/runtime/panic.go:436 +0x32

[2022-07-08T04:47:41.116Z]   github.com/pingcap/tiflow/pkg/causality/internal.TestNodeResolveImmediately()

[2022-07-08T04:47:41.116Z]       /home/jenkins/agent/workspace/cdc_ghpr_leak_test/go/src/github.com/pingcap/tiflow/pkg/causality/internal/node_test.go:121 +0x35

[2022-07-08T04:47:41.116Z]   testing.tRunner()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1439 +0x213

[2022-07-08T04:47:41.116Z]   testing.(*T).Run.func1()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1486 +0x47

[2022-07-08T04:47:41.116Z] 

[2022-07-08T04:47:41.116Z] Previous read at 0x00c0001aac80 by goroutine 7:

[2022-07-08T04:47:41.116Z]   github.com/pingcap/tiflow/pkg/causality/internal.(*Node).Free()

[2022-07-08T04:47:41.116Z]       /home/jenkins/agent/workspace/cdc_ghpr_leak_test/go/src/github.com/pingcap/tiflow/pkg/causality/internal/node.go:90 +0x32

[2022-07-08T04:47:41.116Z]   github.com/pingcap/tiflow/pkg/causality/internal.TestNodeFree.func1()

[2022-07-08T04:47:41.116Z]       /home/jenkins/agent/workspace/cdc_ghpr_leak_test/go/src/github.com/pingcap/tiflow/pkg/causality/internal/node_test.go:36 +0x30

[2022-07-08T04:47:41.116Z]   github.com/stretchr/testify/assert.didPanic()

[2022-07-08T04:47:41.116Z]       /go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/assert/assertions.go:1018 +0xb7

[2022-07-08T04:47:41.116Z]   github.com/stretchr/testify/assert.Panics()

[2022-07-08T04:47:41.116Z]       /go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/assert/assertions.go:1032 +0x84

[2022-07-08T04:47:41.116Z]   github.com/stretchr/testify/require.Panics()

[2022-07-08T04:47:41.116Z]       /go/pkg/mod/github.com/stretchr/testify@v1.7.2-0.20220504104629-106ec21d14df/require/require.go:1622 +0x9c

[2022-07-08T04:47:41.116Z]   github.com/pingcap/tiflow/pkg/causality/internal.TestNodeFree()

[2022-07-08T04:47:41.116Z]       /home/jenkins/agent/workspace/cdc_ghpr_leak_test/go/src/github.com/pingcap/tiflow/pkg/causality/internal/node_test.go:35 +0xc9

[2022-07-08T04:47:41.116Z]   testing.tRunner()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1439 +0x213

[2022-07-08T04:47:41.116Z]   testing.(*T).Run.func1()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1486 +0x47

[2022-07-08T04:47:41.116Z] 

[2022-07-08T04:47:41.116Z] Goroutine 12 (running) created at:

[2022-07-08T04:47:41.116Z]   testing.(*T).Run()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1486 +0x724

[2022-07-08T04:47:41.116Z]   testing.runTests.func1()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1839 +0x99

[2022-07-08T04:47:41.116Z]   testing.tRunner()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1439 +0x213

[2022-07-08T04:47:41.116Z]   testing.runTests()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1837 +0x7e4

[2022-07-08T04:47:41.116Z]   testing.(*M).Run()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1719 +0xa71

[2022-07-08T04:47:41.116Z]   main.main()

[2022-07-08T04:47:41.116Z]       _testmain.go:67 +0x2e4

[2022-07-08T04:47:41.116Z] 

[2022-07-08T04:47:41.116Z] Goroutine 7 (finished) created at:

[2022-07-08T04:47:41.116Z]   testing.(*T).Run()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1486 +0x724

[2022-07-08T04:47:41.116Z]   testing.runTests.func1()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1839 +0x99

[2022-07-08T04:47:41.116Z]   testing.tRunner()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1439 +0x213

[2022-07-08T04:47:41.116Z]   testing.runTests()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1837 +0x7e4

[2022-07-08T04:47:41.116Z]   testing.(*M).Run()

[2022-07-08T04:47:41.116Z]       /usr/local/go/src/testing/testing.go:1719 +0xa71

[2022-07-08T04:47:41.116Z]   main.main()

[2022-07-08T04:47:41.116Z]       _testmain.go:67 +0x2e4

[2022-07-08T04:47:41.116Z] ==================

[2022-07-08T04:47:41.116Z] --- FAIL: TestNodeDependOn (0.00s)

[2022-07-08T04:47:41.116Z] --- FAIL: TestNodeResolveImmediately (0.00s)

[2022-07-08T04:47:41.116Z]     testing.go:1312: race detected during execution of test

[2022-07-08T04:47:41.116Z] --- FAIL: TestSlotsTrivial (0.01s)

[2022-07-08T04:47:41.116Z]     testing.go:1312: race detected during execution of test

[2022-07-08T04:47:41.116Z] FAIL

[2022-07-08T04:47:41.116Z] FAIL	github.com/pingcap/tiflow/pkg/causality/internal	0.027s
```
https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_leak_test/detail/cdc_ghpr_leak_test/30450/pipeline

### Anything else we need to know

- Does this test exist for other branches as well?

- Has there been a high frequency of failure lately?

--- a/pkg/causality/internal/node_test.go
+++ b/pkg/causality/internal/node_test.go
@@ -23,7 +23,9 @@ import (
 var _ Eq[*Node] = &Node{} // Asserts that *Node implements Eq[*Node].
 
 func TestNodeFree(t *testing.T) {
-	t.Parallel()
+	// This case should not be run parallel to
+	// others, for fear that the use-after-free
+	// will race with NewNode() in other cases.
 
 	nodeA := NewNode()
 	nodeA.Free()

